proposals/exception-handling/Exceptions.md

```diff
@@ -143,42 +93,80 @@ Exception indices are used by:
 1. The `throw` instruction which creates a WebAssembly exception with the
    corresponding exception tag, and then throws it.
 
-2. The `br_on_exn` instruction queries an exception to see if it matches the
-   corresponding exception tag denoted by the exception index. If true it
-   branches to the given label and pushes the corresponding argument values of
-   the exception onto the stack.
-
-### The exception reference data type
+2. The `catch` instruction uses the tag to identify if the thrown exception is
+   one it can catch. If true it pushes the corresponding argument values of the
+   exception onto the stack.
 
-Data types are extended to have a new `exnref` type, that refers to an
-exception. The representation of an exception is left to the implementation.
-
-### Try and catch blocks
+### Try-catch blocks
 
 A _try block_ defines a list of instructions that may need to process exceptions
 and/or clean up state when an exception is thrown. Like other higher-level
 constructs, a try block begins with a `try` instruction, and ends with an `end`
-instruction. That is, a try block is sequence of instructions having the
+instruction. That is, a try-catch block is sequence of instructions having the
 following form:
 
 ```
 try blocktype
   instruction*
-catch
+catch i
+  instruction*
+catch j
+  instruction*
+...
+catch_all
   instruction*
 end
 ```
 
-A try block ends with a `catch block` that is defined by the list of
-instructions after the `catch` instruction.
+A try-catch block contains zero or more `catch` blocks and zero or one
+`catch_all` block. All `catch` blocks should precede the `catch_all` block, if
+any. The `catch`/`catch_all` instructions (within the try construct) are called
+the _catching_ instructions. There should be at least `catch` or `catch_all`
```

Comment:
```suggestion
the _catching_ instructions. There should be at least one `catch` or `catch_all`
```

Reply:
What would that mean then? Wouldn't that be just the same as a `block`?

Reply:
I think try-catch is a little different case then just having a list of of something (e.g. functions) that can be empty. It was designed as a two-parts instruction from the beginning, and while I don't think it's groundbreaking change, many parts of toolchain have to treat try without catch as a special case, because most parts are written under the assumption there will be two parts.

Also I think this is a kind of big change that has not been discussed so far, so I think it warrants a separate discussion thread..?

Reply:
There are still two parts, but one is a list, and the list can have any length, including length 0. :) I don't think this is different from any other similar case, where we have deliberately refrained from making length 0 a special case -- there's not a single occurrence of `X^+` in the entire Wasm grammar. :)

Reply:
Yeah, maybe that doesn't need to be treated as a special case. But anyway, I think we should make this a separate issue post first for more visibility than just changing this here and merging it in.